src/base/bittorrent/trackerentry.cpp

```diff
@@ -104,45 +120,54 @@ void TrackerEntry::setTier(const int value)
     m_nativeEntry.tier = value;
 }
 
-int TrackerEntry::numSeeds() const
+int TrackerEntry::numSeeds(const InfoHashFormats hashVersions) const
 {
-    int value = -1;
+    int value = 0;
```

Comment:
Why did you change it? What value does libtorrent return in case of non-working tracker?

Reply:
libtorrent doesn't initialize to some special value, so it is `0`.

Reply:
>Why didn't you add assertions in all methods that accept HashFormats parameter?

Unlike `status()`, here I consider it is reasonable to return `0` when the formats flag are empty.

Reply:
>the formats flag is empty/mismatched (or when there are no endpoints).

But these are completely different cases. "Empty flags" is an invalid parameter, and that's what I talked about above. We should avoid such an invalid use, at least by inserting assertions to catch possible logical errors that can lead to this.

Reply:
The more I look at this problem, the more I wonder: "Why doesn't it just return the result for all available hash formats?". And I'm getting more and more convinced that you don't need to add this controversial parameter, at least in this PR (i.e., until there's some real need for it).

Reply:
>And I'm getting more and more convinced that you don't need to add this controversial parameter (hash formats), at least in this PR (i.e., until there's some real need for it).

Then how are we going to solve #14087 - wrong tracker status?

IMO the `InfoHashFormat` is needed when we start supporting bittorrent 2.0.
Furthermore, it brings a lot of hard questions that I will briefly mention one: currently we use sha1 hash as the hash for a torrent. In bittorrent 2.0, what is the hash for a torrent? sha1 or sha256 or these two hashed together? This question will affect how we identify a torrent as unique or not.

Reply:
>Then how are we going to solve #14087 - wrong tracker status?

As I said above: "Why doesn't it just return the result for all available hash formats?"
I.e. TrackerEntry needs to know about the hash format of its Torrent and return results based on it. If this information is not available through the native `announce_entry` interface, we should pass it when creating TrackerEntry instance.
>the `InfoHashFormat` is needed when we start supporting bittorrent 2.0.

I agree that this type may be required for some purpose, but I still doubt that we will need to use it with TrackerEntry in the way that you are trying to do now (i.e., request information for some specific hash format, generally different from the one Torrent has).

Reply:
>As I said above: "Why doesn't it just return the result for all available hash formats?"

You dodged my question that won't magically solve the problem. Do you realize that the current code in master returns the wrong answer?

>we should pass it when creating TrackerEntry instance.

Then what is the value for `TrackerEntry()` and `TrackerEntry(const QString &url)` constructors?

Reply:
>what is the value for `TrackerEntry()` and `TrackerEntry(const QString &url)` constructors?

In fact, TrackerEntry is not very well designed, but was hastily made as a simple wrapper over a native `announce_entry` class.
It seems that `TrackerEntry()` is only needed to support containers that require the element type to be default constructible. It just creates an invalid instance (it even doesn't have an URL), so it won't be any worse if it has an invalid hash format too.
`TrackerEntry(const QString &url)` is apparently to support some use cases that look incorrect to me. The main purpose of TrackerEntry is to provide some information related to the interaction of the torrent with the tracker (and I would keep it only for this purpose). But it is also used, for example, to represent the trackers of a magnet link, for which just strings are enough, or to add trackers to a torrent, for which something simpler (URL + tier) is enough, because in fact it is not a "tracker entry" if it does not come from the torrent itself and cannot fulfill its purpose.
So I would remake TrackerEntry(and its usage) more thoroughly. However, I can't insist that you do this. So you can continue with this workaround if you want. Maybe I'll do it later, after I finish my current jobs.

Reply:
>If this information (hash format of Torrent) is not available through the native `announce_entry` interface, we should pass it when creating TrackerEntry instance.

I would even just have `torrent_handle` field together with `announce_entry` inside TrackerEntry, so we would be able to query it for hash formats.

Reply:
>In fact, TrackerEntry is not very well designed, but was hastily made as a simple wrapper over a native announce_entry class.

It used to serve well in the era of bittorrent v1, but obviously it became handicapped in era of bittorrent v2... I don't have the motivation/vision to refactor it as the affecting bug isn't very critical/emergent. I'll probably stop here and leave it to you.

>I would even just have torrent_handle field together with announce_entry inside TrackerEntry, so we would be able to query it for hash formats.

I don't mind that, however I want to note that the ability to query only v1 hash statistics from a hybrid torrent (which have both v1, v2 hashes) is a good thing to have, I imagine it will provide more stats/insights to the user.

Reply:
>I want to note that the ability to query only v1 hash statistics from a hybrid torrent (which have both v1, v2 hashes) is a good thing to have

In any case, we will be able to implement it in a more correct way if it is guided from within the TrackerEntry class. We can return data for different hash formats separately (mapped by hash format), but unlike libtorrent, only for those formats that the torrent actually has.

Reply:
>It used to serve well in the era of bittorrent v1, but obviously it became handicapped in era of bittorrent v2

In my comment above I point out BitTorrent independent disadvantages of TrackerEntry.

Reply:
I hope my comments above don't look like unreasonable nitpicks...

Reply:
>I hope my comments above don't look like unreasonable nitpicks...

Well, if you got a better idea then it is reasonable to bring it out. Also I'll close this PR after it is superseded by another.